core/server/common/src/main/java/alluxio/master/journal/raft/UploadObserver.java

```diff
@@ -0,0 +1,166 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.master.journal.raft;
+
+import alluxio.conf.PropertyKey;
+import alluxio.conf.ServerConfiguration;
+import alluxio.exception.status.InvalidArgumentException;
+import alluxio.grpc.DownloadSnapshotPRequest;
+import alluxio.grpc.DownloadSnapshotPResponse;
+import alluxio.grpc.SnapshotData;
+import alluxio.grpc.UploadSnapshotPRequest;
+import alluxio.grpc.UploadSnapshotPResponse;
+
+import com.google.protobuf.UnsafeByteOperations;
+import io.grpc.stub.ClientCallStreamObserver;
+import io.grpc.stub.ClientResponseObserver;
+import io.grpc.stub.StreamObserver;
+import org.apache.commons.io.IOUtils;
+import org.apache.ratis.statemachine.SnapshotInfo;
+import org.apache.ratis.statemachine.impl.SimpleStateMachineStorage;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.function.Function;
+
+/**
+ * A stream observer for uploading a snapshot.
+ *
+ * @param <S> the message type to send
+ * @param <R> the message type to receive
+ */
+public class UploadObserver<S, R>
```

Comment:
Maybe we should rename this to SnapshotUploader, since UploadObserver seems too generic.

Reply:
Renamed.